### PR TITLE
move `_file` to `CTypes` as `c_FILE`

### PR DESCRIPTION
**Pull request by jeremiah-corrado**
This PR implements one of the changes to `openfp` discussed here: https://github.com/chapel-lang/chapel/issues/20146

- The `_file` type was deprecated (kinda). It's a compiler defined type, so I only marked the methods that use it as deprecated for now. No warning is emitted when the type is used directly by user code — maybe this needs to change?
- A new `c_FILE` type was added to the `CTypes` module. 
- openfp and other internal library/package code were updated to use `c_FILE` instead of `_file`. 

In the same discussion, it was decided to deprecate `openfp` and replace it with an initializer on the `file` type. This can't be done until throwing-initializers are implemented.

- [x] paratest passing

### CODE REVIEW

**Review by lydia-duncan — COMMENTED** [review]
What's here looks good, but it's missing a couple of things:

1. a deprecation test for `openfp`
2. I'd expect to see the compiler code supporting `_file` get a deprecation warning generated for it (and a corresponding test).  I think this will involve modifying `dtFile` defined in `compiler/AST/type.cpp` to have the deprecated flag on its TypeSymbol and looking for uses of dtFile in the compiler to see if they're really necessary.  I'm not sure what dyno does about it, if anything at this point.  If that does work, I'd like to see a test with that deprecation warning as well

**Comment by jeremiah-corrado** [comment]
Ok sounds good, I'll see what I can do. Thanks!

**Comment by jeremiah-corrado** [comment]
I went with the approach of removing `_file`s definition from the compiler and replacing it with a deprecated type definition in ChapelBase:
```chapel
type _file = c_FILE;
```

This way, any use of `_file` emits a deprecation warning, but it can still be used interchangeably with `c_FILE`. 

I kept the `deprecated` annotations on functions that used `_file` for documentation purposes. (They are all `pragma "last resort"`ed, so they are never actually called).

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -8768,8 +8768,7 @@ static void moveHaltForUnacceptableTypes(CallExpr* call) {
   } else if (rhsType == dtNil) {
     bool lhsIsPointer = isClassLikeOrPtr(lhsType) ||
                         lhsType == dtCVoidPtr ||
-                        lhsType == dtCFnPtr ||
-                        lhsType == dtFile;
+                        lhsType == dtCFnPtr;
```
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp:8771]: So this is interesting.  I suspect there's not a test of `_file` being explicitly set to `nil`, but it's something I'd expect to continue to work.  It should be a quick test to write, and if it doesn't work, I'd expect the solution to be something like:
1. add a pragma to `compiler/dyno/include/chpl/uast/PragmaList.chpl` specifically for the `c_FILE` type, similar to how `C_PTR_CLASS` has its own pragma today and attach it to `c_FILE`
2. extend `isClassLikeOrPtr` to also check that flag

Comment (jeremiah-corrado) [reply compiler/resolution/functionResolution.cpp:8771]: Looking into this a bit further, I'm not sure I understand the old nilability behavior of `_file`. Currently on main, this fails to compile:
```chapel
var f: _file = nil;
```
with:
```
error: illegal cast from nil to _file
```
which I believe is coming from line 4403 in functionResultion.cpp. 

Were you expecting the above to work for `_file` (and now `c_FILE`)? Or were you referring to  something else like nillable class fields with the `_file` type?

Comment (jeremiah-corrado) [reply compiler/resolution/functionResolution.cpp:8771]: I guess you can't directly assign `nil` to a plain `c_ptr` either (causes `error: Could not coerce 'nil' to 'c_ptr' in initialization`), but you can assign `nil` to an instantiated `c_ptr`:
```chapel
var a: [0..<5] int = 1;
var x : c_ptr = c_ptrTo(a);
x = nil;
```

This still doesn't compile though:
```chapel
extern proc fopen(name: c_string, mode: c_string): _file;
var cf: _file = fopen("./pt.py", "r");
cf = nil;
```
with:

```
error: Cannot assign to _file from nil
```

Comment (lydia-duncan) [reply compiler/resolution/functionResolution.cpp:8771]: Maybe it's specifically for the return value from a function case?

Comment (lydia-duncan) [reply compiler/resolution/functionResolution.cpp:8771]: In any case, if it didn't work on main then I think it's okay to ignore it for now, thanks for looking!

Comment (jeremiah-corrado) [reply compiler/resolution/functionResolution.cpp:8771]: Assigning `nil` to a `_file` via a function that returns `nil` also isn't compiling, so I'm not sure what that was for either.

File: modules/packages/FFTW.chpl [hunk 2]
```diff
@@ -812,8 +812,16 @@ module FFTW {
 
     extern proc fftw_import_wisdom(read_char : fftw_read_char_func, data : c_void_ptr) : c_int;
 
+    extern proc fftw_fprint_plan(p : fftw_plan, ref output_file : c_FILE) : void;
+
+    pragma "last resort"
+    deprecated "the '_file' type is deprecated; use the variant of 'fftw_fprint_plan' that takes a 'c_FILE'"
```
Comment (mppf) [inline modules/packages/FFTW.chpl:818]: but above we have `type _file = c_FILE` ? Maybe you can just remove these and people working with `_file` should get a deprecation wherever that is created?

**Comment by bradcray** [comment]
> This can't be done until throwing-initializers are implemented.

Just to recap some experiments that were done before you joined the team, I believe (and based on my sketchy memory):
* we could do a halting version of the initializer as a step in the right direction
* I _believe_ that we found that if the compiler check for `throws` on an initializer was disabled or ignored, it became possible to write a throwing initializer as long as any throwing was done after the (explicit or implicit) `this.complete()` step, or in the `postinit()` step.  This suggested to me that supporting throwing in those cases alone was not necessarily very far off.

**Pull request merged**